pkg/ddl/schema.go

```diff
@@ -154,6 +161,144 @@ func onModifySchemaDefaultPlacement(jobCtx *jobContext, job *model.Job) (ver int
+			return ver, errors.Trace(err)
+		}
+		job.FinishDBJob(model.JobStateDone, model.StatePublic, ver, dbInfo)
+		return ver, nil
+	}
+	trxTableName := infoschema.ClusterTableTiDBTrx
+	failpoint.Inject("mockModifySchemaReadOnlyDDL", func() {
+		trxTableName = infoschema.TableTiDBTrx
+	})
+
+	switch job.SchemaState {
+	case model.StateNone:
+		dbInfo.ReadOnly = args.ReadOnly
+		if err = jobCtx.metaMut.UpdateDatabase(dbInfo); err != nil {
+			return ver, errors.Trace(err)
+		}
+		if ver, err = updateSchemaVersion(jobCtx, job); err != nil {
+			return ver, errors.Trace(err)
+		}
+		job.SchemaState = model.StatePendingReadOnly
+	case model.StatePendingReadOnly:
+		uncommittedTxn, err := getUncommittedTxnIDs(jobCtx, w.sess, job.Query, dbInfo.ID, trxTableName)
+		failpoint.Inject("mockCheckUncommittedTxnError", func() {
+			if err == nil {
+				err = errors.New("mock error for check uncommitted txn")
+			}
+		})
+		if err != nil {
+			return ver, errors.Trace(err)
+		}
+
+		// check and wait all uncommitted txn
+		for len(uncommittedTxn) > 0 {
+			for txnID := range uncommittedTxn {
+				sql := fmt.Sprintf("SELECT ID FROM INFORMATION_SCHEMA.%s WHERE ID = %d", trxTableName, txnID)
+				r, err := w.sess.Execute(jobCtx.stepCtx, sql, "check if txn committed")
+				if err != nil {
+					return ver, errors.Trace(err)
+				}
+				if len(r) == 0 {
+					delete(uncommittedTxn, txnID)
+					continue
+				}
+				logutil.BgLogger().Info("uncommitted txn block read only ddl", zap.Int64("txn ID", txnID))
+				break
+			}
+			time.Sleep(300 * time.Millisecond)
+			failpoint.InjectCall("checkUncommittedTxns", uncommittedTxn)
+		}
+
+		if err = jobCtx.metaMut.UpdateDatabase(dbInfo); err != nil {
+			return ver, errors.Trace(err)
+		}
+		if ver, err = updateSchemaVersion(jobCtx, job); err != nil {
+			return ver, errors.Trace(err)
+		}
+		job.FinishDBJob(model.JobStateDone, model.StatePublic, ver, dbInfo)
+	}
+	return ver, nil
+}
+
+func getUncommittedTxnIDs(jobCtx *jobContext, sess *sess.Session, ddlQuery string, targetDBID int64, trxTableName string) (map[int64]struct{}, error) {
+	var currTS = uint64(0)
+	err := kv.RunInNewTxn(jobCtx.ctx, jobCtx.store, true, func(_ context.Context, txn kv.Transaction) error {
+		currTS = txn.StartTS()
+		return nil
+	})
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	is := jobCtx.infoCache.GetLatest()
+	sql := fmt.Sprintf(
+		"SELECT RELATED_TABLE_IDS, CURRENT_SQL_DIGEST, ID "+
+			"FROM INFORMATION_SCHEMA.%s "+
+			"WHERE (CURRENT_SQL_DIGEST IS NULL OR CURRENT_SQL_DIGEST != TIDB_ENCODE_SQL_DIGEST('%s')) "+ // exclude ddl
+			"AND ID < %d", // exclude new transactions
+		trxTableName,
+		ddlQuery,
+		currTS,
+	)
+	t := time.Now()
+	r, err := sess.Execute(jobCtx.stepCtx, sql, "get uncommitted txn id")
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	queryDur := time.Since(t)
+	ids := make(map[int64]struct{})
+	for _, row := range r {
+		txnID := row.GetInt64(2)
```

Comment:
```suggestion
		txnID := row.GetInt64(1)
```

After the above comment addressed, `CURRENT_SQL_DIGEST` could be removed in the projection fields (`SELECT RELATED_TABLE_IDS, ID`).